Gems/Atom/RHI/Vulkan/Code/Source/RHI/Image.cpp

```diff
@@ -9,21 +9,376 @@
-#include <RHI/Fence.h>
-#include <RHI/PhysicalDevice.h>
-#include <RHI/MemoryAllocator.h>
-#include <RHI/Device.h>
 #include <RHI/ImageView.h>
+#include <RHI/PhysicalDevice.h>
 #include <RHI/Queue.h>
 #include <RHI/ReleaseContainer.h>
+#include <RHI/StreamingImagePool.h>
+#include <RHI/SwapChain.h>
 
 namespace AZ
 {
     namespace Vulkan
     {
+        // SparseImageInfo functions
+        void SparseImageInfo::Init(const Device& device, VkImage vkImage, const RHI::ImageDescriptor& imageDescriptor)
+        {
+            // Get image memory requirements 
+            VkMemoryRequirements memoryRequirements;
+            device.GetContext().GetImageMemoryRequirements(device.GetNativeDevice(), vkImage, &memoryRequirements);
+            
+            // get sparse memory requirement count;
+            uint32_t sparseMemoryReqsCount = 0;
+            device.GetContext().GetImageSparseMemoryRequirements(device.GetNativeDevice(), vkImage, &sparseMemoryReqsCount, nullptr);
+            AZ_Assert(sparseMemoryReqsCount, "Sparse memory requirements count shouldn't be 0");
+
+            // Get actual requirements
+            AZStd::vector<VkSparseImageMemoryRequirements> sparseImageMemoryRequirements;
+            sparseImageMemoryRequirements.resize(sparseMemoryReqsCount);
+            device.GetContext().GetImageSparseMemoryRequirements(device.GetNativeDevice(), vkImage, &sparseMemoryReqsCount, sparseImageMemoryRequirements.data());
+
+            bool validSparseImage = false;
+            for (const VkSparseImageMemoryRequirements& requirements : sparseImageMemoryRequirements)
+            {
+                if (requirements.formatProperties.aspectMask == VkImageAspectFlagBits::VK_IMAGE_ASPECT_COLOR_BIT)
+                {
+                    m_sparseImageMemoryRequirements = requirements;
+                    validSparseImage = true;
+                    break;
+                }
+            }
+
+            if (!validSparseImage)
+            {
+                AZ_Assert(false, "Non-color sparse image is not supported");
+                return;
+            }
+
+            m_image = vkImage;
+
+            VkExtent3D imageGranularity = m_sparseImageMemoryRequirements.formatProperties.imageGranularity;
+
+            m_blockSizeInBytes = memoryRequirements.alignment;
+
+            m_useSingleMipTail = m_sparseImageMemoryRequirements.formatProperties.flags & VK_SPARSE_IMAGE_FORMAT_SINGLE_MIPTAIL_BIT;
+
+            // calculate block count for one mip tail
+            m_mipTailBlockCount = aznumeric_cast<uint32_t>(m_sparseImageMemoryRequirements.imageMipTailSize/m_blockSizeInBytes);
+            
+            // calculate block count for each non-tail mip levels
+            m_tailStartMip =  aznumeric_cast<uint16_t>(m_sparseImageMemoryRequirements.imageMipTailFirstLod);
+            uint32_t nonTailMipLevels = m_tailStartMip;
+            if (nonTailMipLevels)
+            {
+                m_mipSizes.resize(nonTailMipLevels);
```

Comment:
Does it make sense to combine all this metadata into a struct so you don't need to worry about size synchronization and can reduce the total allocation count?

Reply:
+1. Track one vector instead of many